File: drupal-backend/update-form.js
const axios = require('axios')
const fs = require('fs')

const api_key = fs.readFileSync('../jot_api_key.txt')

function compare(a, b) {
    return a.toLowerCase().localeCompare(b.toLowerCase());
}

async function getNamed(field) {
    const url = 'https://oga-boats.herokuapp.com/'
    const r = await axios.post(url, { query: `query{${field}{name}}`})
    return r.data.data[field].map(a => a.name).filter(a => a != '')
}

async function getSimple(field) {
    url = 'https://oga-boats.herokuapp.com/'
    const r = await axios.post(url, { query: `query{${field}}`})
    return r.data.data[field]
}

async function updateQuestion(form, question, field, newValues) {
    const newValue = newValues.sort(compare).join('|')
    if(question[field] === newValue) {
      console.log('no change', question.name)
      return;
    }
    data = `question[${field}]=${encodeURIComponent(newValue)}`
    console.log(data)
    return axios.post(`http://eu-api.jotform.com/form/${form}/question/${question.qid}?apikey=${api_key}`,
       data,
       { 'Content-Type': 'application/x-www-form-urlencoded' }
    )
}

async function main(form_id) {
    const r = await axios.get(`https://eu-api.jotform.com/form/${form_id}/questions?apikey=${api_key}`)
    questions = r.data.content
    for(qid of Object.keys(questions)) {
        const question = questions[qid]
        let r
        switch(question.name) {
        case 'builder':
            r = await updateQuestion(form_id, question, 'items', await getNamed('builders'))
            break;
        case 'designer':
            r = await updateQuestion(form_id, question, 'items', await getNamed('designers'))
            break;
        case 'className':
            r = await updateQuestion(form_id, question, 'items', await getSimple('classNames'))
            break;
        case 'genericType':
            r = await updateQuestion(form_id, question, 'items', await getSimple('genericTypes'))
            break;
        default: // do nothing
        }
        if(r) console.log(r.data)
    }
}

main('200321541644039')

File: drupal-backend/update-form.py
import requests
import os

api_key = os.environ['JOT_API_KEY']

def sortMapOfNames(l):
    return sorted([o['name'] for o in l if o['name'] != ''], key=str.casefold)

def getNamed(field):
    url = 'https://oga-boats.herokuapp.com/'
    data={"query":'query{'+field+'{name}}'}
    r = requests.post(url, json=data)
    return r.json()["data"][field]

def getSimple(field):
    url = 'https://oga-boats.herokuapp.com/'
    data={"query":'query{'+field+'}'}
    r = requests.post(url, json=data)
    return r.json()["data"][field]

# curl -X POST -d 'question[options]=Crabber 24|Hurd 28|Golant Gaffer' https://eu-api.jotform.com/form/200321541644039/question/56?apikey=$JOT_API_KEY
def updateQuestion(form, qid, field, newValue):
    url = "http://eu-api.jotform.com/form/%s/question/%s?apikey=%s" % (form, qid, api_key)
    data = {"question["+field+"]": newValue}
    r = requests.post(url, data=data)
    print(r.text)
    return r

def main():
    form_id = 200321541644039
    r = requests.get('https://eu-api.jotform.com/user/forms?apikey='+api_key)
    forms = r.json()['content']
    r = requests.get('https://eu-api.jotform.com/form/200321541644039?apikey='+api_key)
    form = r.json()
    r = requests.get('https://eu-api.jotform.com/form/200321541644039/questions?apikey='+api_key)
    questions = r.json()['content']
    bl = getNamed('builders')
    sbl = sortMapOfNames(bl)
    builders = '|'.join(sbl)
    dl = getNamed('designers')
    sdl = sortMapOfNames(dl)
    designers = '|'.join(sdl)
    scl = sorted(getSimple('classNames'), key=str.casefold)
    classNames = '|'.join(scl)
    genericTypes = '|'.join(sorted(getSimple('genericTypes'), key=str.casefold))
    for qid in questions:
        question = questions[qid]
        if question['name'] == 'builder':
            if question['items'] != builders:
                updateQuestion(form_id, qid, 'items', builders)
        elif question['name'] == 'designer':
            if question['items'] != designers:
                updateQuestion(form_id, qid, 'items', designers)
        elif question['name'] == 'className':
            if question['items'] != classNames:
                updateQuestion(form_id, qid, 'items', classNames)
        elif question['name'] == 'genericType':
            if question['items'] != genericTypes:
                updateQuestion(form_id, qid, 'items', genericTypes)

if __name__ == "__main__":
    main()
